File: core/MinimumWindowSubstring/MinimumWindowSubstring.java
package com.dbc;

import java.util.HashMap;
import java.util.Map;

public class MinimumWindowSubstring {
    public String minWindow(String s, String t) {
        Map<Character, Integer> sCounter = new HashMap<>();
        Map<Character, Integer> tCounter = new HashMap<>();
        for (int i = 0; i < t.length(); i++) tCounter.put(t.charAt(i), tCounter.getOrDefault(t.charAt(i), 0) + 1);
        int sTotal = 0, tTotal = t.length();
        int left = 0, right = 0;
        String res = "";
        while (right < s.length()) {
            while (right < s.length() && sTotal < tTotal) {
                if (sCounter.getOrDefault(s.charAt(right), 0) < tCounter.getOrDefault(s.charAt(right), 0)) sTotal++;
                if (tCounter.getOrDefault(s.charAt(right), 0) > 0) sCounter.put(s.charAt(right), sCounter.getOrDefault(s.charAt(right), 0) + 1);
                right++;
            }
            while (left < right && sTotal == tTotal && (tCounter.getOrDefault(s.charAt(left), 0) == 0 || sCounter.getOrDefault(s.charAt(left), 0) > tCounter.getOrDefault(s.charAt(left), 0))) {
                if (sCounter.getOrDefault(s.charAt(left), 0) > tCounter.getOrDefault(s.charAt(left), 0)) sCounter.put(s.charAt(left), sCounter.get(s.charAt(left)) - 1);
                left++;
            }
            if (sTotal == tTotal && (res.equals("") || res.length() > right - left)) res = s.substring(left, right);
            sCounter.put(s.charAt(left), sCounter.getOrDefault(s.charAt(left), 0) - 1);
            left++;
            sTotal--;
        }
        return res;
    }
}

File: core/MinimumWindowSubstring/MinimumWindowSubstring.py
from collections import Counter

def minWindow(s: str, t: str) -> str:
    t_counter, t_total = Counter(t), len(t)
    s_counter, s_total, s_length = Counter(), 0, len(s)
    left, right, res = 0, 0, ""
    while left <= right and right < s_length:
        while right < s_length and s_total < t_total:
            if s_counter[s[right]] < t_counter[s[right]]:
                s_total += 1
            if t_counter[s[right]] > 0:
                s_counter[s[right]] += 1
            right += 1

        while left < right and s_total == t_total and (
                s_counter[s[left]] == 0 or s_counter[s[left]] > t_counter[s[left]]):
            if s_counter[s[left]] > t_counter[s[left]]:
                s_counter[s[left]] -= 1
            left += 1

        if s_total == t_total and (res == "" or len(res) > right - left):
            res = s[left:right]

        s_counter[s[left]] -= 1
        left += 1
        s_total -= 1

    return res

if __name__ == '__main__':
    print(minWindow(s = "ADOBECODEBANC", t = "ABC"))







